server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -651,32 +675,66 @@ public ClusterState execute(BatchExecutionContext<StartedShardUpdateTask> batchE
+                            final Index index = startedShardEntry.shardId.getIndex();
+                            ClusterStateTimeRanges clusterStateTimeRanges = updatedTimestampRanges.get(index);
+                            IndexLongFieldRange currentTimestampMillisRange = clusterStateTimeRanges == null
+                                ? null
+                                : clusterStateTimeRanges.timestampRange();
+                            IndexLongFieldRange currentEventIngestedMillisRange = clusterStateTimeRanges == null
+                                ? null
+                                : clusterStateTimeRanges.eventIngestedRange();
+
                             final IndexMetadata indexMetadata = initialState.metadata().index(index);
                             if (currentTimestampMillisRange == null) {
                                 currentTimestampMillisRange = indexMetadata.getTimestampRange();
                             }
-                            final IndexLongFieldRange newTimestampMillisRange;
-                            newTimestampMillisRange = currentTimestampMillisRange.extendWithShardRange(
-                                entry.shardId.id(),
+                            if (currentEventIngestedMillisRange == null) {
+                                currentEventIngestedMillisRange = indexMetadata.getEventIngestedRange();
+                            }
+
+                            final IndexLongFieldRange newTimestampMillisRange = currentTimestampMillisRange.extendWithShardRange(
+                                startedShardEntry.shardId.id(),
                                 indexMetadata.getNumberOfShards(),
-                                entry.timestampRange
+                                startedShardEntry.timestampRange
                             );
-                            if (newTimestampMillisRange != currentTimestampMillisRange) {
-                                updatedTimestampRanges.put(index, newTimestampMillisRange);
+                            /*
+                             * Only track 'event.ingested' range this if the cluster state min transport version is on/after the version
+                             * where we added 'event.ingested'. If we don't do that, we will have different cluster states on different
+                             * nodes because we can't send this data over the wire to older nodes.
+                             */
+                            IndexLongFieldRange newEventIngestedMillisRange = IndexLongFieldRange.UNKNOWN;
+                            TransportVersion minTransportVersion = batchExecutionContext.initialState().getMinTransportVersion();
+                            if (minTransportVersion.onOrAfter(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)) {
+                                newEventIngestedMillisRange = currentEventIngestedMillisRange.extendWithShardRange(
```

Comment:
I don't think I understand what you are asking. The newly added code around `event.ingested` here mirrors the way the `@timestamp` range is handled a few lines above, with the one modification that the "newRange" gets initialized to UNKNOWN for cases where the min transport version of the cluster is still indicating that not all clusters have upgraded. If all clusters have upgraded to a version that knows how to deal with event.ingested ranges, then we call `extendWithShardRange` off the `currentEventIngestedMillisRange` (not the "newRange") which we get out of index metadata (just like we do for `@timestamp`).

Sorry if I'm missing your concern - can you elaborate on what you think could go wrong here (and how to compensate for it)?

Reply:
We will write UNKNOWN into the index metadata of all indices while in mixed cluster. Any subsequent rounds here will remain on UNKNOWN.

It is more related to how to handle existing indices, which I think you want to do a separate PR for. We could in principle maybe just fix it here, but then it will require all shards to start before the ranges are available, which does not seem nice (and may have other pitfalls). So I think we'll want a separate PR to populate all UNKNOWN entries.

So nothing to do here for now.